adserver/analyzer/tests.py

```diff
@@ -0,0 +1,228 @@
+import datetime
+
+import requests
+import responses
+from django.core.exceptions import ValidationError
+from django.test import TestCase
+from django.utils import timezone
+from django_dynamic_fixture import get
+
+from . import tasks
+from ..models import Offer
+from ..models import Publisher
+from .backends import NaiveKeywordAnalyzerBackend
+from .models import AnalyzedUrl
+from .utils import get_url_analyzer_backend
+from .utils import normalize_url
+from .validators import KeywordsValidator
+
+
+class TestValidators(TestCase):
+    def test_keywords_validator(self):
+        validator = KeywordsValidator(message="Test Message")
+
+        # Ok
+        validator([])
+        validator(["django", "python"])
+        validator(["ruby", "rails", "framework", "web", "http"])
+
+        # Raise Errors
+        self.assertRaises(ValidationError, validator, {})
+        self.assertRaises(ValidationError, validator, "Not-List")
+        self.assertRaises(
+            ValidationError,
+            validator,
+            ["Upper", "Case"],
+        )
+
+
+class TestUtils(TestCase):
+    def test_normalize_url(self):
+        self.assertEqual(
+            normalize_url("https://example.com/path"), "https://example.com/path"
+        )
+        self.assertEqual(
+            normalize_url("https://example.com/path?myparam=myval"),
+            "https://example.com/path?myparam=myval",
+        )
+        self.assertEqual(
+            normalize_url("https://example.com/#fragment"),
+            "https://example.com/#fragment",
+        )
```

Comment:
We presumably want to strip fragments, since we aren't going to take them into account, and they will add multiple URL entries?

Reply:
Without fragments, I think there's no hope for SPAs. I think this may require some experimentation.

Reply:
Ack true... I don't think there's much hope for SPA's anyway, since we aren't using something like headlesschrome to get the URL. That could be a v2, something with JavaScript.